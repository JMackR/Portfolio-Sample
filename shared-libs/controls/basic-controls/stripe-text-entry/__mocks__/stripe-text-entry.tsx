import React from 'react'

export const StripeTextEntry = () => <div>Stripe Text Entry</div>
