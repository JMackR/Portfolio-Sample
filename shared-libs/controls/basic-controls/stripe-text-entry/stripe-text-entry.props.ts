import type { TextColors, TextTypes } from '@sample/themes'

export interface StripeTextEntryProps {
  text?: string

  textType?: keyof TextTypes
  /**
   * Optional override default Stripe element text input color
   * @default: primary1
   */
  textColor?: keyof TextColors
  /**
   * Optional override default Stripe element hint color
   * @default: primary2
   */
  hintColor?: keyof TextColors

  /**
   * Sets the color of the cursor and selection when the text box is in focus.
   */
  tintColor?: keyof TextColors

  /**
   * Text displayed when no content has been entered in the text box.
   */
  hint?: string

  /**
   * Used to provide Stripe element text input
   */
  stripeInputType?: 'card-number' | 'card-exp' | 'card-cvc'
}
