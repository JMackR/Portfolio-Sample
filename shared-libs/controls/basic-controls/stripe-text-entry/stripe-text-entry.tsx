import React from 'react'
import { Flex, Margin } from '@sample/layout-controls'
import { FontStyle, useColorsForTextColorsCollection, useFontForTextType } from '@sample/themes'
import { CardNumberElement, CardExpiryElement, CardCvcElement } from '@stripe/react-stripe-js'
import { StripeTextEntryProps } from './stripe-text-entry.props'

export const StripeTextEntry: React.FC<StripeTextEntryProps> = (props: StripeTextEntryProps) => {
  const font: FontStyle = useFontForTextType(props.textType !== undefined ? props.textType : 'secondary3')
  const [primaryColor, hintColor] = useColorsForTextColorsCollection([
    props.textColor !== undefined ? props.textColor : 'primary1',
    props.hintColor !== undefined ? props.hintColor : 'primary2',
  ])

  const elementForSpecialInputType = () => {
    const stripeStyle = {
      base: {
        color: primaryColor,
        fontWeight: font.fontWeight,
        fontFamily: `${font.fontFamily}, sans-serif`,
        fontSize: `${font.fontSize}px`,
        '::placeholder': {
          color: hintColor,
        },
      },
      invalid: {
        fontWeight: font.fontWeight,
        fontFamily: `${font.fontFamily}, sans-serif`,
        fontSize: `${font.fontSize}px`,
      },
    }

    const stripeOptions = {
      style: stripeStyle,
      placeholder: props.hint,
    }

    switch (props.stripeInputType) {
      case 'card-number':
        return <CardNumberElement options={stripeOptions} />
      case 'card-exp':
        return <CardExpiryElement options={stripeOptions} />
      case 'card-cvc':
        return <CardCvcElement options={stripeOptions} />
      default: {
        return <CardNumberElement options={stripeOptions} />
      }
    }
  }

  return (
    <Margin marginTopStep={0.5}>
      <Flex direction={'column'} grow={1}>
        {elementForSpecialInputType()}
      </Flex>
    </Margin>
  )
}
