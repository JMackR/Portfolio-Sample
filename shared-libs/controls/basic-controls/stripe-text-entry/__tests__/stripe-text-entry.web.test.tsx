import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { StripeTextEntry } from '../stripe-text-entry'
import { StripeTextEntryProps } from '../stripe-text-entry.props'

jest.mock('@stripe/react-stripe-js', () => ({
  CardNumberElement: () => <div>CardNumberElement</div>,
  CardExpiryElement: () => <div>CardExpiryElement</div>,
  CardCvcElement: () => <div>CardCvcElement</div>,
}))

describe('Stripe Text Entry', () => {
  let props: StripeTextEntryProps

  beforeEach(() => {
    props = {}
  })

  it('should render the CardNumberElement by default', () => {
    const { getByText } = render(<StripeTextEntry {...props} />)

    expect(getByText('CardNumberElement')).toBeDefined()
  })

  describe('stripeInputType is card-number', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-number'
    })

    it('should render the CardNumberElement', () => {
      const { getByText } = render(<StripeTextEntry {...props} />)

      expect(getByText('CardNumberElement')).toBeDefined()
    })
  })

  describe('stripeInputType is card-exp', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-exp'
    })

    it('should render the CardExpiryElement', () => {
      const { getByText } = render(<StripeTextEntry {...props} />)

      expect(getByText('CardExpiryElement')).toBeDefined()
    })
  })

  describe('stripeInputType is card-cvc', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-cvc'
    })

    it('should render the CardCvcElement', () => {
      const { getByText } = render(<StripeTextEntry {...props} />)

      expect(getByText('CardCvcElement')).toBeDefined()
    })
  })
})

describe('Stripe Text Entry Snapshot Tests', () => {
  let props: StripeTextEntryProps
  beforeEach(() => {
    props = {}
  })

  it('should render the CardNumberElement by default', () => {
    const tree = renderer.create(<StripeTextEntry {...props} />)

    expect(tree).toMatchSnapshot()
  })

  describe('stripeInputType is card-number', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-number'
    })

    it('should render the CardNumberElement', () => {
      const tree = renderer.create(<StripeTextEntry {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('stripeInputType is card-exp', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-exp'
    })

    it('should render the CardExpiryElement', () => {
      const tree = renderer.create(<StripeTextEntry {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('stripeInputType is card-cvc', () => {
    beforeEach(() => {
      props.stripeInputType = 'card-cvc'
    })

    it('should render the CardCvcElement', () => {
      const tree = renderer.create(<StripeTextEntry {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('text/hint color and textType are passed in', () => {
    beforeEach(() => {
      props.textColor = 'error1'
      props.hintColor = 'error1'
      props.textType = 'heading5'
    })

    it('should render the CardCvcElement', () => {
      const tree = renderer.create(<StripeTextEntry {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })
})
