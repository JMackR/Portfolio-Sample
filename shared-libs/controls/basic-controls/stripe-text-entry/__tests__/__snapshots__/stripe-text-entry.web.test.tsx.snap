// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stripe Text Entry Snapshot Tests should render the CardNumberElement by default 1`] = `
.margin {
  margin: undefined;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: 2px;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.flex {
  display: flex;
  flex-basis: auto;
  flex-grow: 1;
  flex-shrink: 1;
  flex-direction: column;
  flex-wrap: undefined;
  align-items: stretch;
  justify-content: flex-start;
  max-width: 100vw;
  width: undefined;
  height: undefined;
  position: relative;
  background-color: undefined;
  overflow: undefined;
}

<div
  className="margin"
>
  <div
    className="flex"
  >
    <div>
      CardNumberElement
    </div>
  </div>
</div>
`;

exports[`Stripe Text Entry Snapshot Tests stripeInputType is card-cvc should render the CardCvcElement 1`] = `
.margin {
  margin: undefined;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: 2px;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.flex {
  display: flex;
  flex-basis: auto;
  flex-grow: 1;
  flex-shrink: 1;
  flex-direction: column;
  flex-wrap: undefined;
  align-items: stretch;
  justify-content: flex-start;
  max-width: 100vw;
  width: undefined;
  height: undefined;
  position: relative;
  background-color: undefined;
  overflow: undefined;
}

<div
  className="margin"
>
  <div
    className="flex"
  >
    <div>
      CardCvcElement
    </div>
  </div>
</div>
`;

exports[`Stripe Text Entry Snapshot Tests stripeInputType is card-exp should render the CardExpiryElement 1`] = `
.margin {
  margin: undefined;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: 2px;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.flex {
  display: flex;
  flex-basis: auto;
  flex-grow: 1;
  flex-shrink: 1;
  flex-direction: column;
  flex-wrap: undefined;
  align-items: stretch;
  justify-content: flex-start;
  max-width: 100vw;
  width: undefined;
  height: undefined;
  position: relative;
  background-color: undefined;
  overflow: undefined;
}

<div
  className="margin"
>
  <div
    className="flex"
  >
    <div>
      CardExpiryElement
    </div>
  </div>
</div>
`;

exports[`Stripe Text Entry Snapshot Tests stripeInputType is card-number should render the CardNumberElement 1`] = `
.margin {
  margin: undefined;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: 2px;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.flex {
  display: flex;
  flex-basis: auto;
  flex-grow: 1;
  flex-shrink: 1;
  flex-direction: column;
  flex-wrap: undefined;
  align-items: stretch;
  justify-content: flex-start;
  max-width: 100vw;
  width: undefined;
  height: undefined;
  position: relative;
  background-color: undefined;
  overflow: undefined;
}

<div
  className="margin"
>
  <div
    className="flex"
  >
    <div>
      CardNumberElement
    </div>
  </div>
</div>
`;

exports[`Stripe Text Entry Snapshot Tests text/hint color and textType are passed in should render the CardCvcElement 1`] = `
.margin {
  margin: undefined;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: 2px;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.flex {
  display: flex;
  flex-basis: auto;
  flex-grow: 1;
  flex-shrink: 1;
  flex-direction: column;
  flex-wrap: undefined;
  align-items: stretch;
  justify-content: flex-start;
  max-width: 100vw;
  width: undefined;
  height: undefined;
  position: relative;
  background-color: undefined;
  overflow: undefined;
}

<div
  className="margin"
>
  <div
    className="flex"
  >
    <div>
      CardNumberElement
    </div>
  </div>
</div>
`;
